OpenDreamRuntime/Procs/Native/DreamProcNativeSavefile.cs

```diff
@@ -9,29 +13,124 @@ internal static class DreamProcNativeSavefile {
-            throw new NotImplementedException("General support for ExportText() is not implemented");
+        string result = ExportTextInternal(savefile);
+
+        savefile.CurrentPath = oldPath; //restore current directory after a query
+        if(!file.IsNull){
+            if(file.TryGetValueAsString(out var fileStr)) {
+                File.WriteAllText(fileStr, result);
+            } else if(file.TryGetValueAsDreamResource(out var fileResource)) {
+                fileResource.Output(new DreamValue(result));
+            } else {
+                throw new ArgumentException($"Invalid file value {file}");
+            }
+        }
+        return new DreamValue(result);
+    }
+
+    private static string ExportTextInternal(DreamObjectSavefile savefile, int indent = int.MinValue) {
+        string result = "";
+        var value = savefile.CurrentDir;
+        var key = savefile.CurrentPath.Split('/').Last();
+        if(indent == int.MinValue){
+            key = ".";
+            indent = 0; //either way, set indent to 0 so we know we're not at the start anymore
+        }
+        switch(value) {
+            case DreamObjectSavefile.SFDreamPrimitive primitiveValue:
+                if(primitiveValue.Value.IsNull)
+                    result += $"{new string('\t', indent)}{key} = null\n";
+                else switch(primitiveValue.Value.Type){
+                    case DreamValue.DreamValueType.String:
+                        result += $"{new string('\t', indent)}{key} = \"{primitiveValue.Value.MustGetValueAsString()}\"\n";
+                        break;
+                    case DreamValue.DreamValueType.Float:
+                        result += $"{new string('\t', indent)}{key} = {primitiveValue.Value.MustGetValueAsFloat()}\n";
+                        break;
+                }
+                break;
+            case DreamObjectSavefile.SFDreamFileValue fileValue:
+                result += $"{new string('\t', indent)}{key} = \nfiledata(\"";
+                result += $"name={fileValue.Name};";
+                result += $"ext={fileValue.Ext};";
+                result += $"length={fileValue.Length};";
+                result += $"crc32=0x{fileValue.Crc32:x8};";
+                result += $"encoding=base64\",{{\"\n{fileValue.Data}\n\"}}";
+                result += ")\n";
+                break;
+            case DreamObjectSavefile.SFDreamObjectPathValue objectValue:
+                result += $"{new string('\t', indent)}{key} = object(\"{objectValue.Path}\")\n";
+                break;
+            case DreamObjectSavefile.SFDreamType typeValue:
+                result += $"{new string('\t', indent)}{key} = {typeValue.TypePath}\n";
+                break;
+            case DreamObjectSavefile.SFDreamListValue listValue:
+                result += $"{new string('\t', indent)}{key} = {ExportTextInternalListFormat(listValue)}\n";
+                break;
+            case DreamObjectSavefile.SFDreamDir jsonValue:
```

Comment:
`jsonValue` is unused
```suggestion
            case DreamObjectSavefile.SFDreamDir:
```